pkg/binlog/position.go

```diff
@@ -18,22 +18,96 @@ import (
 	"strings"
 
 	"github.com/pingcap/errors"
-	"github.com/siddontang/go-mysql/mysql"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+
+	"github.com/pingcap/dm/pkg/utils"
+)
+
+const (
+	// in order to differ binlog pos from multi (switched) masters, we added a UUID-suffix field into binlogPos.Name
+	// and we also need support: with UUIDSuffix's pos should always > without UUIDSuffix's pos, so we can update from @without to @with automatically
+	// conversion: originalPos.NamePrefix + posUUIDSuffixSeparator + UUIDSuffix + binlogFilenameSep + originalPos.NameSuffix => convertedPos.Name
+	// UUIDSuffix is the suffix of sub relay directory name, and when new sub directory created, UUIDSuffix is incremented
+	// eg. mysql-bin.000003 in c6ae5afe-c7a3-11e8-a19d-0242ac130006.000002 => mysql-bin|000002.000003
+	// where `000002` in `c6ae5afe-c7a3-11e8-a19d-0242ac130006.000002` is the UUIDSuffix
+	posUUIDSuffixSeparator = "|"
 )
 
 // PositionFromStr constructs a mysql.Position from a string representation like `mysql-bin.000001:2345`
-func PositionFromStr(s string) (mysql.Position, error) {
+func PositionFromStr(s string) (gmysql.Position, error) {
 	parsed := strings.Split(s, ":")
 	if len(parsed) != 2 {
-		return mysql.Position{}, errors.New("the format should be filename:pos")
+		return gmysql.Position{}, errors.New("the format should be filename:pos")
 	}
 	pos, err := strconv.ParseUint(parsed[1], 10, 32)
 	if err != nil {
-		return mysql.Position{}, errors.New("the pos should be digital")
+		return gmysql.Position{}, errors.New("the pos should be digital")
 	}
 
-	return mysql.Position{
+	return gmysql.Position{
 		Name: parsed[0],
 		Pos:  uint32(pos),
 	}, nil
 }
+
+// RealMySQLPos parses a relay position and returns a mysql position and whether error occurs
+// if parsed successfully and `UUIDSuffix` exists, sets position Name to
+// `originalPos.NamePrefix + binlogFilenameSep + originalPos.NameSuffix`.
+// if parsed failed returns the given position and the traced error.
+func RealMySQLPos(pos gmysql.Position) (gmysql.Position, error) {
+	parsed, err := ParseFilename(pos.Name)
+	if err != nil {
+		return pos, errors.Trace(err)
+	}
+
+	sepIdx := strings.Index(parsed.BaseName, posUUIDSuffixSeparator)
+	if sepIdx > 0 && sepIdx+len(posUUIDSuffixSeparator) < len(parsed.BaseName) {
+		return gmysql.Position{
+			Name: ConstructFilename(parsed.BaseName[:sepIdx], parsed.Seq),
+			Pos:  pos.Pos,
+		}, nil
+	}
+
+	return pos, nil
+}
+
+// ExtractPos extracts (uuidWithSuffix, uuidSuffix, originalPos) from input pos (originalPos or convertedPos)
+func ExtractPos(pos gmysql.Position, uuids []string) (uuidWithSuffix string, uuidSuffix string, realPos gmysql.Position, err error) {
+	if len(uuids) == 0 {
+		err = errors.NotValidf("empty UUIDs")
+		return
+	}
+
+	parsed, _ := ParseFilename(pos.Name)
```

Comment:
don't need handle the err?

Reply:
done in https://github.com/pingcap/dm/pull/140/commits/cf17487d1339271972b9846d950214c23e88e13a